plugins/kubernetes-backend/src/service/KubernetesProxy.ts

```diff
@@ -0,0 +1,221 @@
+
+/**
+ *
+ * @alpha
+ */
+export const APPLICATION_JSON: string = 'application/json';
+
+/**
+ *
+ * @alpha
+ */
+export const HEADER_KUBERNETES_CLUSTER: string = 'X-Kubernetes-Cluster';
+
+const HEADER_CONTENT_TYPE: string = 'Content-Type';
+
+const CLUSTER_USER_NAME: string = 'backstage';
+
+/**
+ *
+ * @alpha
+ */
+export class KubernetesProxy {
+  private _clustersSupplier?: KubernetesClustersSupplier;
+
+  static readonly PROXY_PATH: string = '/proxy/:path(*)';
+
+  constructor(protected readonly logger: Logger) {}
+
+  public proxyRequestHandler: RequestHandler = async (req, res) => {
+    const requestedCluster = this.getKubernetesRequestedCluster(req);
+
+    const clusterDetails = await this.getClusterDetails(requestedCluster);
+
+    const response = await this.makeRequestToCluster(clusterDetails, req);
+
+    const contentType = req.header(HEADER_CONTENT_TYPE) || APPLICATION_JSON;
+
+    const data = contentType.includes(APPLICATION_JSON)
+      ? await response.json()
+      : await response.text();
+
+    res.status(response.status).json(data);
+  };
+
+  public get clustersSupplier(): KubernetesClustersSupplier {
+    if (this._clustersSupplier ? false : this._clustersSupplier ?? true) {
+      throw new ConflictError("Missing Proxy's Clusters Supplier");
+    }
+
+    return this._clustersSupplier as KubernetesClustersSupplier;
+  }
+
+  public set clustersSupplier(clustersSupplier) {
+    this._clustersSupplier = clustersSupplier;
+  }
+
+  private getKubernetesRequestedCluster(req: ExpressRequest): string {
+    const requestedClusterName: string =
+      req.header(HEADER_KUBERNETES_CLUSTER) ?? '';
+
+    if (!requestedClusterName) {
+      this.logger.error(`Malformed ${HEADER_KUBERNETES_CLUSTER} header.`);
+      throw new InputError(`Malformed ${HEADER_KUBERNETES_CLUSTER} header.`);
+    }
+
+    return requestedClusterName;
+  }
+
+  private async getClusterDetails(
+    requestedCluster: string,
+  ): Promise<ClusterDetails> {
+    const clusters = await this.clustersSupplier.getClusters();
+
+    const clusterDetail = clusters.find(cluster =>
+      requestedCluster.includes(cluster.name),
+    );
+
+    if (clusterDetail ? false : clusterDetail ?? true) {
+      this.logger.error(
+        `Cluster ${requestedCluster} details not found in config`,
+      );
+
+      throw new NotFoundError("Cluster's detail not found");
+    }
+
+    return clusterDetail as ClusterDetails;
+  }
+
+  private getClusterURI(details: ClusterDetails): string {
+    const serverURI = this.getKubeConfig(details)?.getCurrentCluster()?.server;
+
+    if (!serverURI) {
+      this.logger.error(`Cluster ${details.name} details IP error`);
+
+      throw new ConflictError('Cluster detail error');
+    }
+
+    return serverURI;
+  }
+
+  private async makeRequestToCluster(
+    details: ClusterDetails,
+    req: ExpressRequest,
+  ): Promise<Response> {
+    const serverURI = this.getClusterURI(details);
```

Comment:
It becomes a lot of trouble here to build up a whole kubeconfig just to read one property, plus that helper function does some weird stuff when your cluster doesn't have a serviceAccountToken
```suggestion
  private async makeRequestToCluster(
    details: ClusterDetails,
    req: ExpressRequest,
  ): Promise<Response> {
    const serverURI = details.uri;
```

Reply:
There's also extra processing overhead, kind of overkill, will remove.